File: python-probes/src/soma_probes/flax/v1/modules/encoder.py
from jax import Array
from flax import nnx
from soma_probes.config import (
    V1_NUM_LAYERS,
    V1_EMBEDDING_DIM,
    V1_PWFF_HIDDEN_DIM,
    V1_NUM_HEADS,
    V1_MAX_WAVELENGTH,
)
from soma_probes.flax.v1.modules.layer import Layer, LayerConfig
from dataclasses import dataclass


@dataclass
class EncoderConfig:
    dropout_rate: float
    embedding_dim: int = V1_EMBEDDING_DIM
    pwff_hidden_dim: int = V1_PWFF_HIDDEN_DIM
    num_layers: int = V1_NUM_LAYERS
    num_heads: int = V1_NUM_HEADS
    max_wavelength: int = V1_MAX_WAVELENGTH


class Encoder(nnx.Module):
    def __init__(self, config: EncoderConfig, rngs: nnx.Rngs):
        self.layers = nnx.List(
            [
                Layer(
                    LayerConfig(
                        dropout_rate=config.dropout_rate,
                        embedding_dim=config.embedding_dim,
                        pwff_hidden_dim=config.pwff_hidden_dim,
                        num_heads=config.num_heads,
                        max_wavelength=config.max_wavelength,
                    ),
                    rngs=rngs,
                )
                for _ in range(config.num_layers)
            ]
        )

    def __call__(
        self,
        representations: Array,
        positions: Array,
    ):
        x = representations
        for layer in self.layers:
            x = layer(x, positions)
        return x

File: probes/src/v1/modules/encoder.rs
use burn::{
    config::Config,
    module::Module,
    nn::Initializer,
    tensor::{Int, Tensor, backend::Backend},
};

use crate::v1::{V1_EMBEDDING_DIM, V1_NUM_LAYERS, V1_PWFF_HIDDEN_DIM, modules::layer::Layer};

#[derive(Config, Debug)]
pub struct EncoderConfig {
    /// The size of the input and output features.
    #[config(default = "V1_EMBEDDING_DIM")]
    pub embedding_dim: usize,
    /// The size of the hidden inner pwff features.
    #[config(default = "V1_PWFF_HIDDEN_DIM")]
    pub pwff_hidden_dim: usize,
    /// The number of transformer layers.
    #[config(default = "V1_NUM_LAYERS")]
    pub num_layers: usize,
    /// The probability that dropout occurs
    #[config(default = 0.0)]
    pub dropout_rate: f64,
    /// The type of function used to initialize neural network parameters
    #[config(
        default = "Initializer::KaimingUniform{gain:1.0/num_traits::Float::sqrt(3.0), fan_out_only:false}"
    )]
    pub initializer: Initializer,
}

#[derive(Module, Debug)]
pub struct Encoder<B: Backend> {
    layers: Vec<Layer<B>>,
}

impl EncoderConfig {
    pub fn init<B: Backend>(&self, device: &B::Device) -> Encoder<B> {
        let layers = (0..V1_NUM_LAYERS)
            .map(|_| Layer::<B>::new(self, device))
            .collect::<Vec<_>>();
        Encoder { layers }
    }
}
impl<B: Backend> Encoder<B> {
    pub fn forward(
        &self,
        representations: Tensor<B, 3>,
        positions: Tensor<B, 2, Int>,
    ) -> Tensor<B, 3> {
        let mut x = representations;
        for layer in self.layers.iter() {
            x = layer.forward(x, positions.clone());
        }
        x
    }
}
